domain/machine/state/state_test.go

```diff
@@ -74,6 +74,45 @@ func (s *stateSuite) TestCreateMachine(c *gc.C) {
 	c.Assert(machineID, gc.Equals, "666")
 }
 
+// TestCreateMachineAlreadyExists asserts that a MachineAlreadyExists error is
+// returned when the machine already exists.
+func (s *stateSuite) TestCreateMachineAlreadyExists(c *gc.C) {
+	err := s.state.CreateMachine(context.Background(), "666", "", "")
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.state.CreateMachine(context.Background(), "666", "", "")
+	c.Check(err, jc.ErrorIs, machineerrors.MachineAlreadyExists)
+}
+
+// TestCreateMachineWithParentSuccess asserts the happy path of
+// CreateMachineWithParent at the state layer.
+func (s *stateSuite) TestCreateMachineWithParentSuccess(c *gc.C) {
+	// Create the parent first
+	err := s.state.CreateMachine(context.Background(), "666", "3", "1")
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Create the machine with the created parent
+	err = s.state.CreateMachineWithParent(context.Background(), "667", "666", "4", "2")
+	c.Check(err, jc.ErrorIsNil)
```

Comment:
**todo:** I don't think it's enough to check that the error is nil. We have to check that the machine has been truly created in db.

If we don't, imagine: someone break the code and the machine is silently not created.

- This test pass (it doesn't check the error)
- The test `TestCreateMachineAlreadyExists` fails (yes, because the arrange part doesn't create a machine so the error `machineAlreadyExists` is never returned. 

Test should fails when what they test broke. We should focus on designing test to break in a way that analysis would be straightforward.

**thought:** Test design should be focused on the way they break, not how to make them pass ;)

Reply:
No need to check that in every test, `TestCreateMachine` already manually checks if a machine has been truly created. `TestCreateMachineWithParentSuccess` should only be about the logic of creating a machine with a parent.

Reply:
I agree on the fact that it should only be about the logic of creating a machine with parent, but the expected output is not "non nil" error, the expected output of this function is a side effect, which contains two part:

* a machine has been created (no need to check all their attribute, which can be covered by a simple "createMachineTest"
* a link between this machine and its parent has been created.

None of those side effect have been tested, which leads me to my comment.

Reply:
@gfouillet done, see 41b8669fb95fbd5967c5c607262e3a8dbbc6fc85